code/campaignclassic/src/main/java/com/adobe/marketing/mobile/AEPPushTemplateBroadcastReceiver.java

```diff
@@ -0,0 +1,302 @@
+                if (cachedLeftImage != null) {
+                    cachedImages.add(BitmapFactory.decodeStream(cachedLeftImage.getData()));
+                }
+            }
+
+            if (!StringUtils.isNullOrEmpty(centerImageUri)) {
+                final CacheResult cachedCenterImage =
+                        cacheService.get(assetCacheLocation, centerImageUri);
+                if (cachedCenterImage != null) {
+                    cachedImages.add(BitmapFactory.decodeStream(cachedCenterImage.getData()));
+                }
+            }
+
+            if (!StringUtils.isNullOrEmpty(rightImageUri)) {
+                final CacheResult cachedRightImage =
+                        cacheService.get(assetCacheLocation, rightImageUri);
+                if (cachedRightImage != null) {
+                    cachedImages.add(BitmapFactory.decodeStream(cachedRightImage.getData()));
+                }
+            }
+        }
+
+        final RemoteViews smallLayout =
+                new RemoteViews(packageName, R.layout.push_template_collapsed);
+        final RemoteViews expandedLayout =
+                new RemoteViews(packageName, R.layout.push_template_filmstrip_carousel);
+        smallLayout.setTextViewText(R.id.notification_title, titleText);
+        smallLayout.setTextViewText(R.id.notification_body, smallBodyText);
+        expandedLayout.setTextViewText(R.id.notification_title, titleText);
+        expandedLayout.setTextViewText(R.id.notification_body_expanded, expandedBodyText);
+
+        final String action = intent.getAction();
+        String newLeftCaption;
+        String newCenterCaption;
+        String newRightCaption;
+        Bitmap newLeftImage;
+        Bitmap newCenterImage;
+        Bitmap newRightImage;
+        String newLeftImageUri;
+        String newCenterImageUri;
+        String newRightImageUri;
+
+        if (action.equals(CampaignPushConstants.IntentActions.FILMSTRIP_LEFT_CLICKED)) {
+            newCenterImage = cachedImages.get(0);
+            newRightImage = cachedImages.get(1);
+            newLeftImage = cachedImages.get(2);
+            newCenterImageUri = leftImageUri;
+            newRightImageUri = centerImageUri;
+            newLeftImageUri = rightImageUri;
+            newLeftCaption = captionMap.get(rightImageUri);
+            newCenterCaption = captionMap.get(leftImageUri);
+            newRightCaption = captionMap.get(centerImageUri);
+        } else {
+            newCenterImage = cachedImages.get(2);
+            newRightImage = cachedImages.get(0);
+            newLeftImage = cachedImages.get(1);
+            newCenterImageUri = rightImageUri;
+            newRightImageUri = leftImageUri;
+            newLeftImageUri = centerImageUri;
+            newLeftCaption = captionMap.get(centerImageUri);
+            newCenterCaption = captionMap.get(rightImageUri);
+            newRightCaption = captionMap.get(leftImageUri);
+        }
```

Comment:
Given that we support more than 3 images in a carousel/filmstrip/auto carousel, I don't think this logic will not work for templates with more than 3 images as this cycles through `0 1 2` and `1 2 0` indices explicitly.
In order to expand this for multiple images (with the constraint that we cannot store/cache this carousel items map data structure efficiently with the existing file based cache used by CacheService), I think we should pass 
1. all the item image urls, item captions, click actions for the images as individual String arrays with the intent.
2. the index of the current image at the center.

Then when you receive the intent on a click, we can identify the next & previous item based on the current index and use the data from the images.

Reply:
makes sense, i'll update this